src/doc/tarpl/references.md

```diff
@@ -0,0 +1,150 @@
+% References
+
+There are two kinds of reference:
+
+* Shared reference: `&`
+* Mutable reference: `&mut`
+
+Which obey the following rules:
+
+* A reference cannot outlive its referent
+* A mutable reference cannot be aliased
+
+To define aliasing, we must define the notion of *paths* and *liveness*.
+
+
+
+
+# Paths
+
+If all Rust had were values, then every value would be uniquely owned
+by a variable or composite structure. From this we naturally derive a *tree*
+of ownership. The stack itself is the root of the tree, with every variable
+as its direct children. Each variable's direct children would be their fields
+(if any), and so on.
+
+From this view, every value in Rust has a unique *path* in the tree of ownership.
+References to a value can subsequently be interpreted as a path in this tree.
+Of particular interest are *ancestors* and *descendants*: if `x` owns `y`, then
+`x` is an *ancestor* of `y`, and `y` is a *descendant* of `x`. Note that this is
+an inclusive relationship: `x` is a descendant and ancestor of itself.
+
+Tragically, plenty of data doesn't reside on the stack, and we must also accommodate this.
+Globals and thread-locals are simple enough to model as residing at the bottom
+of the stack (though we must be careful with mutable globals). Data on
+the heap poses a different problem.
+
+If all Rust had on the heap was data uniquely owned by a pointer on the stack,
+then we can just treat that pointer as a struct that owns the value on
+the heap. Box, Vec, String, and HashMap, are examples of types which uniquely
+own data on the heap.
+
+Unfortunately, data on the heap is not *always* uniquely owned. Rc for instance
+introduces a notion of *shared* ownership. Shared ownership means there is no
+unique path. A value with no unique path limits what we can do with it. In general, only
+shared references can be created to these values. However mechanisms which ensure
+mutual exclusion may establish One True Owner temporarily, establishing a unique path
+to that value (and therefore all its children).
+
+The most common way to establish such a path is through *interior mutability*,
+in contrast to the *inherited mutability* that everything in Rust normally uses.
+Cell, RefCell, Mutex, and RWLock are all examples of interior mutability types. These
+types provide exclusive access through runtime restrictions. However it is also
+possible to establish unique ownership without interior mutability. For instance,
+if an Rc has refcount 1, then it is safe to mutate or move its internals.
+
+In order to correctly communicate to the type system that a variable or field of
```

Comment:
nit: I don't think we're communicating to the type system, we're communicating with the optimizer that this isn't regular single-owned data. I think.

The only communication with the type system I see is that `UnsafeCell` is `!Sync/Send` by default.